From b76c5dc970ddc8aa562393f580637e2c1b5b6ed7 Mon Sep 17 00:00:00 2001
From: Antoine Cotten <hello@acotten.com>
Date: Thu, 14 Apr 2022 17:55:13 +0200
Subject: [PATCH] Switch from 'pci_' to 'dma_' API in Linux >= 5.18

The legacy pci_dma_ API was removed in Linux 5.18.

See also: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=d56baf6efaf14e2910610216c581ca71d6940012
          https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7968778914e53788a01c2dee2692cab157de9ac0

Original patch by Joan Bruguera:
https://gist.github.com/joanbm/052d8e951ba63d5eb5b6960bfe4e031a
---
 src/shared/linux_osl.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/src/shared/linux_osl.c b/src/shared/linux_osl.c
index 3970dca..3d99002 100644
--- a/src/shared/linux_osl.c
+++ b/src/shared/linux_osl.c
@@ -48,7 +48,11 @@ typedef struct bcm_mem_link {
 struct osl_info {
 	osl_pubinfo_t pub;
 	uint magic;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 	void *pdev;
+#else
+	struct pci_dev *pdev;
+#endif
 	atomic_t malloced;
 	atomic_t pktalloced; 	
 	uint failed;
@@ -600,7 +604,11 @@ osl_dma_alloc_consistent(osl_t *osh, uint size, uint16 align_bits, uint *alloced
 	if (va)
 		*pap = (ulong)__virt_to_phys(va);
 #else
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 	va = pci_alloc_consistent(osh->pdev, size, (dma_addr_t*)pap);
+#else
+	va = dma_alloc_coherent(&osh->pdev->dev, size, (dma_addr_t *)pap, GFP_ATOMIC);
+#endif
 #endif
 	return va;
 }
@@ -613,7 +621,11 @@ osl_dma_free_consistent(osl_t *osh, void *va, uint size, ulong pa)
 #ifdef __ARM_ARCH_7A__
 	kfree(va);
 #else
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 	pci_free_consistent(osh->pdev, size, va, (dma_addr_t)pa);
+#else
+	dma_free_coherent(&osh->pdev->dev, size, va, (dma_addr_t)pa);
+#endif
 #endif
 }
 
@@ -623,7 +635,11 @@ osl_dma_map(osl_t *osh, void *va, uint size, int direction, void *p, hnddma_seg_
 	int dir;
 
 	ASSERT((osh && (osh->magic == OS_HANDLE_MAGIC)));
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 	dir = (direction == DMA_TX)? PCI_DMA_TODEVICE: PCI_DMA_FROMDEVICE;
+#else
+	dir = (direction == DMA_TX)? DMA_TO_DEVICE : DMA_FROM_DEVICE;
+#endif
 
 #if defined(__ARM_ARCH_7A__) && defined(BCMDMASGLISTOSL)
 	if (dmah != NULL) {
@@ -635,13 +651,21 @@ osl_dma_map(osl_t *osh, void *va, uint size, int direction, void *p, hnddma_seg_
 			if (skb_is_nonlinear(skb)) {
 				nsegs = skb_to_sgvec(skb, sg, 0, PKTLEN(osh, skb));
 				ASSERT((nsegs > 0) && (totsegs + nsegs <= MAX_DMA_SEGS));
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 				pci_map_sg(osh->pdev, sg, nsegs, dir);
+#else
+				dma_map_sg(&osh->pdev->dev, sg, nsegs, dir);
+#endif
 			} else {
 				nsegs = 1;
 				ASSERT(totsegs + nsegs <= MAX_DMA_SEGS);
 				sg->page_link = 0;
 				sg_set_buf(sg, PKTDATA(osh, skb), PKTLEN(osh, skb));
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 				pci_map_single(osh->pdev, PKTDATA(osh, skb), PKTLEN(osh, skb), dir);
+#else
+				dma_map_single(&osh->pdev->dev, PKTDATA(osh, skb), PKTLEN(osh, skb), dir);
+#endif
 			}
 			totsegs += nsegs;
 			totlen += PKTLEN(osh, skb);
@@ -656,7 +680,11 @@ osl_dma_map(osl_t *osh, void *va, uint size, int direction, void *p, hnddma_seg_
 	}
 #endif 
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 	return (pci_map_single(osh->pdev, va, size, dir));
+#else
+	return (dma_map_single(&osh->pdev->dev, va, size, dir));
+#endif
 }
 
 void BCMFASTPATH
@@ -665,8 +693,13 @@ osl_dma_unmap(osl_t *osh, uint pa, uint size, int direction)
 	int dir;
 
 	ASSERT((osh && (osh->magic == OS_HANDLE_MAGIC)));
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 0)
 	dir = (direction == DMA_TX)? PCI_DMA_TODEVICE: PCI_DMA_FROMDEVICE;
 	pci_unmap_single(osh->pdev, (uint32)pa, size, dir);
+#else
+	dir = (direction == DMA_TX)? DMA_TO_DEVICE : DMA_FROM_DEVICE;
+	dma_unmap_single(&osh->pdev->dev, (uint32)pa, size, dir);
+#endif
 }
 
 #if defined(BCMDBG_ASSERT)
